master/internal/stream/supervisor.go

```diff
@@ -0,0 +1,150 @@
+package stream
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gorilla/websocket"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/syncx/errgroupx"
+)
+
+const (
+	minInterval = time.Second
+	maxInterval = time.Minute
+	maxRetries  = 5
+)
+
+// Supervisor manages the context for underlying PublisherSet.
+type Supervisor struct {
+	cond *sync.Cond
+	// these things change each restart of the PublisherSet
+	subctx context.Context
+	ps     *PublisherSet
+}
+
+// NewSupervisor creates a new Supervisor.
+func NewSupervisor() *Supervisor {
+	var lock sync.Mutex
+	return &Supervisor{cond: sync.NewCond(&lock)}
+}
+
+// monitorPermissionChanges listens for permission changes, updates the PublisherSet
+// to signal to boot streamers, returns an error in the event of a failure to listen.
+func monitorPermissionChanges(ctx context.Context, ps *PublisherSet) error {
+	permListener, err := AuthZProvider.Get().GetPermissionChangeListener()
+	if err != nil {
+		log.Errorf("unable to get permission change listener: %s", err)
+		// XXX: investigate better recovery mechanism
+		return err
+	}
+	if permListener == nil {
+		// no listener means we don't have permissions configured at all
+		return nil
+	}
+	defer func() {
+		err := permListener.Close()
+		if err != nil {
+			log.Debugf("error occurred while closing permission listener: %s", err)
+		}
+	}()
+
+	for {
+		select {
+		// did permissions change?
+		case <-permListener.Notify:
+			log.Debugf("permission change detected, booting streamers")
+			func() {
+				ps.permissionLock.Lock()
+				defer ps.permissionLock.Unlock()
+				close(ps.permissionChangeChan)
+				ps.permissionChangeChan = make(chan struct{})
```

Comment:
This should be a method on the PublisherSet.

The idea of separating the monitorPermissionsChanges from the PublihserSet is to give PublisherSet a method that implements the just bootem logic (I'd actually call it "BootEm" since that's the name we use in the DESIGN.md doc), that you can call from unit tests.

But if the monitorPermissionChanges is just locking the PublisherSet, modifying its members, and unlocking it, then you don't really ahve any separation at all.

Reply:
I like this idea a lot! This way the PublisherSet synchronization is being handled by the PublisherSet, instead of pushing the responsibility on anyone who's using it